pyccel/ast/core.py

```diff
@@ -2537,6 +2577,108 @@ def __str__(self):
     def is_unused(self):
         return False
 
+class InlineFunctionDef(FunctionDef):
+    """
+    Represents a function definition for an inline function.
+
+    Parameters
+    ----------
+    See FunctionDef
+
+    namespace_imports : Scope
+                        The objects in the scope which are available due to imports
+    """
+    __slots__ = ('_namespace_imports','_orig_args','_new_args','_new_local_vars', '_if_block_replacements')
+
+    def __init__(self, *args, namespace_imports = None, **kwargs):
+        self._namespace_imports = namespace_imports
+        super().__init__(*args, **kwargs)
+        self._orig_args = tuple(a.var for a in self.arguments)
+        self._new_args  = None
+        self._new_local_vars = None
+        self._if_block_replacements = None
+
+    @property
+    def is_inline(self):
+        """ True if the function should be printed inline """
+        return True
+
+    @property
+    def namespace_imports(self):
+        """ The objects in the scope which are available due to imports
+        """
+        return self._namespace_imports
+
+    def swap_in_args(self, args, new_local_vars):
+        """ Modify the body of the function by replacing the argument and local Variables
+        with the provided argument and local Variables
+        """
+        assert self._new_args is None
+        assert self._new_local_vars is None
+        # Collect the function arguments and the expressions they will be replaced with
+        self._new_local_vars = tuple(new_local_vars)
+
+        # We cannot replace with singletons as this cannot be reversed
+        self._new_args  = tuple(NilArgument() if a is Nil() else \
+                        PyccelAssociativeParenthesis(a) if isinstance(a, PyccelOperator) \
+                        else a for a in args)
+
+        # Replace the arguments in the code
+        self.body.substitute(self._orig_args+self.local_vars, self._new_args+self._new_local_vars, invalidate=False)
+
+    def swap_out_args(self):
+        """ Modify the body of the function by reinstating the original argument and local Variables
+        """
+        self.body.substitute(self._new_args+self._new_local_vars, self._orig_args+self.local_vars, invalidate=False)
+        self._new_args = None
+        self._new_local_vars = None
+
+    def remove_presence_checks(self):
+        """ Modify the body by replacing all expressions checking for the presence of an optional
+        variable. Either the If is removed or the check is replaced with its literal result
+        """
```

Comment:
I am not sure why you think this function would be useful in `FunctionDef` so I am going to describe in more detail what it does.

The aim of this function is to replace expressions such as:
```
if a is None:
    x = 1
else:
    x = 2
```
with (for example) `x = 1.
This function is called after `swap_in_args` when `PyccelIs` objects can be evaluated.

`FunctionDef` doesn't have this function as it does not handle inline situations so the result of the `PyccelIs` statements is always unknown

Reply:
The aim is for:
```
@inline
def get_val(x = None : float):
    if x is None:
        a = 0
    else:
        a = int(x)
    return a

def g():
    a = get_val(3.5)
    b = get_val(29.4)
    c = get_val()
```

to translate to:
```
def g():
    a_0001 = int(3.5)
    a = a_0001
    a_0001 = int(29.4)
    b = a_0001
    a_0001 = 0
    c = a_0001
```
instead of:
```
def g():
    if 3.5 is None:
        a_0001 = 0
    else:
        a_0001 = int(3.5)
    a = a_0001
    if 29.4 is None:
        a_0001 = 0
    else:
        a_0001 = int(29.4)
    b = a_0001
    if None is None:
        a_0001 = 0
    else:
        a_0001 = None
    c = a_0001
```
The latter would be problematic as several of these statements can't be printed in c/fortran

Reply:
I see. I thought this was an optimization step, instead it is used because you cannot use Fortran `present` once you inline the function...

Reply:
But then my I was not too far off when I asked the following
> Or maybe it's just a matter of not being able to print some function-specific instructions when in-lining the function body?

Reply:
Indeed.
`present(2)`
Does not make sense and `a = None` cannot be printed